lib/web/integrations.go

```diff
@@ -0,0 +1,185 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/web/ui"
+)
+
+// integrationsCreate creates an Integration
+func (h *Handler) integrationsCreate(w http.ResponseWriter, r *http.Request, p httprouter.Params, sctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	var req *ui.Integration
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var ig *types.IntegrationV1
+	var err error
+
+	switch req.SubKind {
+	case types.IntegrationSubKindAWSOIDC:
+		ig, err = types.NewIntegrationAWSOIDC(
+			types.Metadata{Name: req.Name},
+			&types.AWSOIDCIntegrationSpecV1{
+				RoleARN: req.AWSOIDC.RoleARN,
+			},
+		)
+
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+	default:
+		return nil, trace.BadParameter("subkind %q is not supported", req.SubKind)
+	}
+
+	clt, err := sctx.GetUserClient(r.Context(), site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	storedIntegration, err := clt.CreateIntegration(r.Context(), ig)
+	if err != nil {
+		if trace.IsAlreadyExists(err) {
+			return nil, trace.AlreadyExists("failed to create Integration (%q already exists), please use another name", req.Name)
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	return ui.MakeIntegration(storedIntegration), nil
+}
+
+// integrationsUpdate updates the Integration based on its name
+func (h *Handler) integrationsUpdate(w http.ResponseWriter, r *http.Request, p httprouter.Params, sctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	integrationName := p.ByName("name")
+	if integrationName == "" {
+		return nil, trace.BadParameter("an integration name is required")
+	}
+
+	var req *ui.UpdateIntegrationRequest
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clt, err := sctx.GetUserClient(r.Context(), site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	integration, err := clt.GetIntegration(r.Context(), integrationName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if req.AWSOIDC != nil {
+		if integration.GetSubKind() != types.IntegrationSubKindAWSOIDC {
+			return nil, trace.BadParameter("cannot update %q fields for a %q integration", types.IntegrationSubKindAWSOIDC, integration.GetSubKind())
+		}
```

Comment:
Can this logic be encapsulated into a `CanUpdate()` method on the integration resource? So it can be used by other clients as well, e.g. tctl will also need it, right? Check out a similar method on the trusted cluster resource.

Reply:
I was able to apply that pattern to `tctl` but not here

We are using a type specific to Web.
To re-use the `CanChangeStateTo` here, we would need to re-create the integration to then pass it to the existing integration's `CanChangeStateTo`.

I think we should keep the `switch case` here.

Let me know what you think :pray: